lib/events/joinrulesevent.h

```diff
@@ -0,0 +1,18 @@
+// SPDX-FileCopyrightText: 2021 Carl Schwan <carl@carlschwan.eu>
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#pragma once
+
+#include "stateevent.h"
+namespace Quotient {
+class JoinRulesEvent : public StateEvent {
+public:
+    QUO_EVENT(JoinRulesEvent, "m.room.join_rules")
+
+    explicit JoinRulesEvent(const QJsonObject& obj) : StateEvent(obj) {}
+
+    QString joinRule() const;
+    QJsonArray allow() const;
+};
```

Comment:
This is a rather basic implementation of `m.room.join_rules`, missing the safeguards that I would expect it to have. At the very least, `joinRule` should be an enum (hopefully we don't get a tide of new join rules in the next year, as we've just added a bunch). Even better, `allow` could actually be made a vector of `AllowCondition` structures [defined in the spec](https://spec.matrix.org/v1.5/client-server-api/#mroomjoin_rules). I realise this is more work than just copying code from NeoChat - if you don't feel like doing it, let me know and I will help.